### PR TITLE
Change the FF_MakeNameCompliant behavior

### PR DESCRIPTION
**Pull request by togrue**
* Change the handling of invalid file-names to cause an error. Depending on the entry to create (directory/file), a different error code is returned.

Description
-----------
This changes the behavior of the FF_CreateDirent function to return an error in case of an invalid path. 
Filenames with the first character of 0xE5 are still modified. (Which might have some issues. TODO)

Note: Invalid directory paths and invalid file paths now result in two different FreeRTOS+ return values. Is this the intended behavior or should only one error code be used?
```
case FF_ERR_FILE_INVALID_PATH:
    return pdFREERTOS_ERRNO_ENOTDIR; /* The path of the file was not found. */
case FF_ERR_DIR_INVALID_PATH:
    return pdFREERTOS_ERRNO_EINVAL; /* Could not find the directory specified by the path. */
```

FF_MakeNameCompliant returns a boolean and not an error-code, because the function
would need an additional parameter to select the right error code (FF_ERR_DIR_INVALID_PATH or FF_ERR_FILE_INVALID_PATH).

Test Steps
-----------
A invalid filepath "test:file.txt" resulted in setting the FF_ERR_FILE_INVALID_PATH. (Tested through breakpoint in prvFFErrorToErrno).

Todo:
- mkdir with invalid path
- handling of names starting with 0xE5

Checklist:
----------

- [ ] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by paulbartell** [comment]
@togrue Thanks for your contribution. To get this merged, could you run uncrustify on your changes and address the other CI failures (spell check and build check)?

File: ff_dir.c [hunk 1]
```diff
@@ -2886,7 +2886,7 @@ FF_Error_t FF_ExtendDirectory( FF_IOManager_t * pxIOManager,
 /* Windows says: don't use these characters: '\/:*?"<>|'
  *  "     *     /	    :     <     >     ?    '\'    ?     | */
     0x22, 0x2A, 0x2F, 0x3A, 0x3C, 0x3E, 0x3F, 0x5C, 0x7F, 0x7C
-};
+}
```
Comment (togrue) [inline ff_dir.c:2889]: Uncrustify removed the semicolon. 
I agree, it looks nicer without it, but now it doesn't build.

**Comment by togrue** [comment]
Sorry for the force-push, i thought i can modify my fork, without changing the pull-request.

**Comment by Skptak** [comment]
Thanks for raising this PR, and no issue with the force push. @cookpate @phelter how do these changes look to you?

**Comment by htibosch** [comment]
One moment please, I am working on it

File: ff_dir.c [hunk 2]
```diff
@@ -2941,160 +2949,165 @@ FF_Error_t FF_CreateDirent( FF_IOManager_t * pxIOManager,
...
-    {
-        xEntryCount = 1;
-    }
-    #endif
-
-    /* Create the ShortName. */
-    FF_LockDirectory( pxIOManager );
-
-    do
+    if( FF_MakeNameCompliant( pxDirEntry->pcFileName ) )
```
Comment (htibosch) [inline ff_dir.c:2952]: The checker for MISRA rules will prefer to see a comparison here:
~~~diff
-    if( FF_MakeNameCompliant( pxDirEntry->pcFileName ) )
+    if( FF_MakeNameCompliant( pxDirEntry->pcFileName ) == pdTRUE )
~~~

File: ff_dir.c [hunk 3]
```diff
@@ -2941,160 +2949,165 @@ FF_Error_t FF_CreateDirent( FF_IOManager_t * pxIOManager,
...
+            if( pxDirEntry != NULL )
+            {
+                pxDirEntry->usCurrentItem = ( uint16_t ) ( lFreeEntry + xLFNCount );
+            }
+        }
+    } /* if( FF_MakeNameCompliant( pxDirEntry->pcFileName ) ) */
+    else
     {
-        if( pxDirEntry != NULL )
+        if( pxDirEntry->ucAttrib & FF_FAT_ATTR_DIR )
```
Comment (htibosch) [inline ff_dir.c:3104]: Very good, return a proper error code for tracing.

**Review by htibosch — APPROVED** [review]
You chose for a different approach than I would have done. But yours is good too: you just refuse to process an invalid file name at the lowest level: at `FF_CreateDirent()`.
I approve this PR.

File: ff_dir.c (hunk 3 above)
Comment (htibosch) [inline ff_dir.c:3104]: Sorry one more remark related to MISRA:

~~~diff
-    if( pxDirEntry->ucAttrib & FF_FAT_ATTR_DIR )
+     if( ( pxDirEntry->ucAttrib & FF_FAT_ATTR_DIR ) != 0U )
~~~
This test (`!= 0U`) will make it a real boolean value.

File: ff_dir.c [hunk 4]
```diff
@@ -135,9 +135,9 @@ static BaseType_t FF_ValidShortChar( char cChar );
 #endif /* if ( ffconfigLFN_SUPPORT != 0 ) */
 
 #if ( ffconfigUNICODE_UTF16_SUPPORT != 0 )
-    static void FF_MakeNameCompliant( FF_T_WCHAR * pcName );
+    static FF_Error_t FF_MakeNameCompliant( FF_T_WCHAR * pcName );
 #else
-    static void FF_MakeNameCompliant( char * pcName );
+    static FF_Error_t FF_MakeNameCompliant( char * pcName );
```
[the remark this answers is not in the document]
Comment (htibosch) [reply ff_dir.c:140]: > So please change it to :
> 
> ```
>     static BaseType_t FF_MakeNameCompliant( char * pcName );
> ```
Or even change the name:
~~~diff
-    static BaseType_t FF_MakeNameCompliant( char * pcName );
+    static BaseType_t FF_CheckNameCompliance( char * pcName );
~~~
because it will not change the name anymore.

But once again, thanks for your issue report #61 and thank you for the solution!

File: ff_dir.c [hunk 5]
```diff
@@ -2875,24 +2875,31 @@ FF_Error_t FF_ExtendDirectory( FF_IOManager_t * pxIOManager,
 } /* FF_ExtendDirectory() */
 /*-----------------------------------------------------------*/
 
-static const uint8_t forbiddenChrs[] =
+/* *INDENT-OFF* */
```
Comment (phelter) [inline ff_dir.c:2878]: Doesn't look like  INDENT-OFF is doing anything here.

File: ff_dir.c [hunk 6]
```diff
@@ -2875,24 +2875,31 @@ FF_Error_t FF_ExtendDirectory( FF_IOManager_t * pxIOManager,
...
 #if ( ffconfigUNICODE_UTF16_SUPPORT != 0 )
-    static void FF_MakeNameCompliant( FF_T_WCHAR * pcName )
+    static FF_Error_t FF_MakeNameCompliant( FF_T_WCHAR * pcName )
 #else
-    static void FF_MakeNameCompliant( char * pcName )
+    static FF_Error_t FF_MakeNameCompliant( char * pcName )
 #endif
 /* *INDENT-ON* */
 {
+    FF_Error_t xReturn = pdTRUE;
```
Comment (phelter) [inline ff_dir.c:2899]: Also suggest not having an xReturn just use the fixed value in your `return` statement.  Only use xReturn if you need to decide on the return value and still perform some actions.  - i.e. prefer early returns in your code but I don't know the full `coding guidelines` for this repo so this style might not be preferred.

```suggestion
FF_Error_t xReturn = FF_ERR_NONE;
```

File: ff_dir.c [hunk 7]
```diff
@@ -2903,11 +2910,12 @@ static const uint8_t forbiddenChrs[] =
         {
             if( *pcName == forbiddenChrs[ index ] )
             {
-                *pcName = '_';
-                break;
+                xReturn = pdFALSE;
```
Comment (phelter) [inline ff_dir.c:2913]: Might want to look at: [Error codes](https://github.com/FreeRTOS/Lab-Project-FreeRTOS-FAT/blob/bfb03575759a18e54f1ae3c1dbcb464741463589/include/ff_error.h#L174) and select a more appropriate one.
```suggestion
                xReturn = FF_ERR_FILE_COULD_NOT_CREATE_DIRENT;
```

File: ff_dir.c (hunk 2 above)
Comment (phelter) [reply ff_dir.c:2952]: Use the ErrorCode names not pdTRUE.

```suggestion
    if( FF_MakeNameCompliant( pxDirEntry->pcFileName )  == FF_ERR_NONE)
```

**Review by phelter — CHANGES_REQUESTED** [review]

File: ff_dir.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (togrue) [reply ff_dir.c:140]: What about the case of the first filename character being '\xE5'?
The character has to be remapped to 0x05 to have no collision with the file deleted marker (which is 0xE5).

My tests showed that, **filenames starting with '\xE5' still result in duplicated file-entries.**

To handle this case the function `FF_MakeNameCompliant` has to change the filename.

File: ff_dir.c (hunk 3 above)
Comment (togrue) [reply ff_dir.c]: Okay

File: ff_dir.c (hunk 6 above)
Comment (togrue) [reply ff_dir.c]: I would prefer an early return in this function too.
Is it allowed?

File: ff_dir.c (hunk 7 above)
Comment (htibosch) [reply ff_dir.c:2913]: @phelter wrote:
> Might want to look at: [Error codes](https://github.com/FreeRTOS/Lab-Project-FreeRTOS-FAT/blob/bfb03575759a18e54f1ae3c1dbcb464741463589/include/ff_error.h#L174) and select a more appropriate one.

I do not see any reason to use an `FF_Error_t` here because the function is only checking the validity of a string. No disk access is needed.

**Review by htibosch — COMMENTED** [review]
You wrote:

> What about the case of the first filename character being '\xE5'?
> The character has to be remapped to 0x05 to have no collision with the file deleted marker (which is 0xE5).

> My tests showed that, filenames starting with '\xE5' still result in duplicated file-entries.

Yes, it escaped from my attention.

If we want to keep this feature, it must be checked in 3 functions:
~~~
FF_MkDir()
FF_Move()
FF_CreateFile()
~~~

And then we can as well have `FF_MakeNameCompliant()` return void, as it used to do. And preserve the same behaviour as before.

**Comment by htibosch** [comment]
Hi @togrue, I worked out a different approach, please see this attachment:

[make_compliant.zip](https://github.com/FreeRTOS/Lab-Project-FreeRTOS-FAT/files/14217966/make_compliant.zip)

`FF_CheckNameCompliant()` will become a global function, and it will be called as early as possible in three locations:
~~~
FF_CreateFile()
FF_MkDir()
FF_Move()
~~~
That avoids that FAT changes are made and removed, in case `FF_CheckNameCompliant()` failed.

Also it repairs this loop:

~~~diff
#if ( ffconfigUNICODE_UTF16_SUPPORT != 0 )
-    for( index = 0; index < ( BaseType_t ) sizeof( forbiddenChars ); index++ )
+    for( index = 0; index < ( BaseType_t ) ( sizeof( forbiddenChars ) / sizeof ( forbiddenChars[ 0 ] ) ); index++ )
~~~
Without that the UTF16 version would not be correct.

File: ff_dir.c (hunk 6 above)
Comment (htibosch) [reply ff_dir.c]: Early returns are not allowed by the MISRA rules. Sometimes you see this construction:
~~~c
void foo()
{
    FF_Error_t xReturn = FF_ERR_NONE;

    /* This while loop is only introduced to be able to use break. */
    do
    {
        xError = FF_CreateDirent();
        if( FF_isERR( xReturn ) == pdFALSE )
        {
            break;
        }
    }
    while( pdFALSE );

    return xReturn ;
}
~~~
So in a way, this also looks like an early return.

The advantage: when debugging, you can put a break on the one and only return statement.

**Comment by htibosch** [comment]
It looks good, thank you. I am testing all changes when using `FF_CreateFile()`, `FF_MkDir()`, and `FF_Move()`. The last function needs another change. I'll report back here.

**Comment by htibosch** [comment]
The last version of `ff_move()` would always fail because we were checking the compliance of a whole path including subdirectories, like `/etc/file.txt`, where only `file.txt` must be checked.

Please have a look at this version of [`ff_move()`](https://github.com/FreeRTOS/Lab-Project-FreeRTOS-FAT/files/14236760/ff_move.zip)

The tests with `FF_CreateFile()` and `FF_MkDir()` went all well.

**Comment by togrue** [comment]
> The last version of `ff_move()` would always fail because we were checking the compliance of a whole path including subdirectories, like `/etc/file.txt`, where only `file.txt` must be checked.
> 
> The tests with `FF_CreateFile()` and `FF_MkDir()` went all well.

Is there a testsuite, or does everyone have to rewrite all tests?

I have a failing testcase with 
open(".secret.txt", "w") and mkdir(".secretdir") 
It can create duplicated entries, when called multiple times. (However that should be part of another issue?)

Screenshot 2024-02-13 014513

**Comment by htibosch** [comment]
One more proposal for ff_file.c: 

~~~diff
-    if( uxIndex == 0U )
-    {
-        /* Give the directory part a minimum
-         * length of 1, to get '/' at least. */
-        uxIndex = 1U;
-    }
 
     /* Copy the base name of the destination file. */
     STRNCPY( xMyFile.pcFileName, ( szDestinationFile + uxIndex + 1 ), ffconfigMAX_FILENAME - 1 );
     xMyFile.pcFileName[ ffconfigMAX_FILENAME - 1 ] = 0;
 
     /* Now check if the target base name is compliant. */
     if( FF_IsNameCompliant( xMyFile.pcFileName ) == pdFALSE )
     {
         xError = FF_createERR( FF_ERR_FILE_INVALID_PATH, FF_MOVE );
     }

+    if( uxIndex == 0U )
+    {
+        /* Give the directory part a minimum
+         * length of 1, to get '/' at least. */
+        uxIndex = 1U;
+    }
~~~

@togrue wrote:
> Is there a testsuite, or does everyone have to rewrite all tests?

Yes there are, for instance [ff_stdio_tests_with_cwd.c](https://github.com/rjvo/storage/blob/master/ROjal_MQTT_temp/FreeRTOS_example/FreeRTOS-Plus/Demo/Common/FreeRTOS_Plus_FAT_Demos/test/ff_stdio_tests_with_cwd.c).

Another good test that I often do is FTP copying the entire repo of Linux from a laptop to the DUT and copy it back.
After that I compare the results on my laptop, and also I put the SD-card of the DUT in a Linux machine to have it tested on correctness.

Here you find the tests that I ran today:

[ff_tests.zip](https://github.com/FreeRTOS/Lab-Project-FreeRTOS-FAT/files/14257622/ff_tests.zip)

File: ff_dir.c (hunk 5 above)
Comment (togrue) [reply ff_dir.c:2872]: No. It is needed.

**Comment by htibosch** [comment]
@togrue, you wrote:
> I have a failing testcase with
> open(".secret.txt", "w") and mkdir(".secretdir")
> It can create duplicated entries, when called multiple times

Have you re-tested this after the [last change](https://github.com/FreeRTOS/Lab-Project-FreeRTOS-FAT/pull/63#issuecomment-1940617301)?

**Comment by togrue** [comment]
> Have you re-tested this after the https://github.com/FreeRTOS/Lab-Project-FreeRTOS-FAT/pull/63#issuecomment-1940617301?

I used too long filenames without enabling LFN support. 
I fear, that creating files with too long names (for the short-filename entries) should actually lead to an error.
But this is not part of this pull-request.

**Comment by htibosch** [comment]
> I fear, that creating files with too long names (for the short-filename entries) should actually lead to an error.

I must admit that I have LFN enabled in all of my projects that use +FAT.
It would be worth retesting the +FAT library while LFN is disabled.

Another option to re-test would be Unicode/UTF16.

**Comment by htibosch** [comment]
Can anybody please have a look to get this PR merged?
I think it is well tested and ready to merge.
Thanks, Hein

**Review by htibosch — APPROVED** [review]
To summarise the PR: when an application wants to create a file with illegal characters, the file or directory will **not** be created.

In the current version, the illegal characters are being replaced with underscores, which was not a good idea.

Create `"file:01.txt"` results in `"file_01.txt"`.
But when you want to updated `"file:01.txt"`, the fill will not be found and a second instance is created.

So thank you @togrue for this PR. ONce again I approve it.

**Review by AniruddhaKanhere — APPROVED** [review]
